updater.go

```diff
@@ -244,19 +245,108 @@ func (u *Updater) AppendHeaderAt(fh *FileHeader, offset int64) (io.Writer, error
 		return nil, err
 	}
 
-	if offset < 0 {
-		offset = u.dirOffset
-	}
 	// Offset should match existing header offsets or equal to directory offset.
 	var offsetExists bool
-	for i, h := range u.dir {
-		if h.offset == uint64(offset) {
-			offsetExists = true
-			// Delete the corresponding header.
-			u.dir = append(u.dir[:i], u.dir[i+1:]...)
-			break
+
+	existingFileIndex := 0
+	if offset < 0 {
+		// Append the file to the zip or replace the existing one
+		for i, h := range u.dir {
+			if h.FileHeader.Name == fh.Name {
+				offset = int64(h.offset)
+				existingFileIndex = i
+				break
+			}
+		}
+
+		// File does not exist inside the zip, so we append it
+		if existingFileIndex == 0 {
+			offset = u.dirOffset
+		}
+	} else {
+		// This is for manually writing data into a certain offset
+		for i, h := range u.dir {
+			if h.offset == uint64(offset) {
+				offsetExists = true
+				// Delete the corresponding header.
+				u.dir = append(u.dir[:i], u.dir[i+1:]...)
+				break
+			}
 		}
 	}
+
+	if existingFileIndex > 0 {
+		// If the file exists and is not the last one in the file, we will remove it
+		// and reinsert it at the end of the file.
+		// The existing files data will be relocated first to use the leftover space.
+
+		// This diagram shows the internal structure of a zip file, whose i-th file
+		// we want to overwrite. The annotations represent some of the variables that
+		// are used to implement this feature.
+		//
+		//                       deletedDataSize = nextOffset - existingFileOffset
+		//                      ┌────────────────┐
+		//                      │                │ remainingDataSize
+		//                      │                │────────────────────┐
+		//                      ▼                ▼                    ▼
+		// ┌───┬────────────┬───┬───┬────────────┬───┬────────────┬───┬───────┐
+		// │   │            │   │   │            │   │            │   │░░░░░░░│
+		// │h_0│   File 0   │...│h_i│   File i   │h_ │  File i+1  │...│░░Dir░░│
+		// │   │    data    │   │   │    data    │i+1│    data    │   │░░░░░░░│
+		// │   │            │   │   │            │   │            │   │░░░░░░░│
+		// └───┴────────────┴───┴───┴────────────┴───┴────────────┴───┴───────┘
+		//                      ▲                ▲                    ▲
+		//                      │                │                    │
+		//                   existingFileOffset  nextOffset           Dir Offset
+		if existingFileIndex+1 < len(u.dir) {
+			existingFileOffset := u.dir[existingFileIndex].offset
+			nextOffset := u.dir[existingFileIndex+1].offset
+
+			// Read the existing files data
+			remainingDataSize := u.dirOffset - int64(nextOffset)
+			data := make([]byte, remainingDataSize)
+			_, err := u.rw.ReadAt(data, int64(nextOffset))
+			if err != nil {
+				return nil, err
+			}
+
+			// Rewind the ReadWriter offset to the one of the file to be deleted
+			_, err = u.rw.Seek(int64(u.dir[existingFileIndex].offset), io.SeekStart)
```

Comment:
The `u.dir[existingFileIndex].offset` can be replaced with `existingFileOffset`.

Reply:
Oops, this slipped during me refactoring and creating that variable.